storage/storage_image.go

```diff
@@ -464,15 +464,29 @@ func (s *storageImageDestination) TryReusingBlob(ctx context.Context, blobinfo t
 		}, nil
 	}
 
-	// Does the blob correspond to a known DiffID which we already have available?
-	// Because we must return the size, which is unknown for unavailable compressed blobs, the returned BlobInfo refers to the
-	// uncompressed layer, and that can happen only if canSubstitute.
-	if canSubstitute {
-		if uncompressedDigest := cache.UncompressedDigest(blobinfo.Digest); uncompressedDigest != "" && uncompressedDigest != blobinfo.Digest {
-			layers, err := s.imageRef.transport.store.LayersByUncompressedDigest(uncompressedDigest)
-			if err != nil && errors.Cause(err) != storage.ErrLayerUnknown {
-				return false, types.BlobInfo{}, errors.Wrapf(err, `Error looking for layers with digest %q`, uncompressedDigest)
+	// Does the blob correspond to a compressed version of a known DiffID which we have available?
+	if uncompressedDigest := cache.UncompressedDigest(blobinfo.Digest); uncompressedDigest != "" && uncompressedDigest != blobinfo.Digest {
+		layers, err := s.imageRef.transport.store.LayersByUncompressedDigest(uncompressedDigest)
+		if err != nil && errors.Cause(err) != storage.ErrLayerUnknown {
+			return false, types.BlobInfo{}, errors.Wrapf(err, `Error looking for layers with digest %q`, uncompressedDigest)
+		}
+		// If one of the layers with the matching diffID was initialized using a compressed
+		// blob that matches the input, and we recorded the size of that blob, we can report
+		// that digest+size pair.
+		for _, layer := range layers {
+			if layer.CompressedDigest == blobinfo.Digest {
```

Comment:
How can this ever be reached?  If `layer.CompressedDigest == blobinfo.Digest`, `s.imageRef.transport.store.LayersByCompressedDigest(blobinfo.Digest)` above, around line 453, should have returned a non-empty list.

Reply:
Hmm, you're right, that wouldn't give us the size of the compressed version of the layer.  I think we should throw this one out.